src/zkevm_specs/evm/execution/end_block.py

```diff
@@ -1,28 +1,74 @@
+# which can be achieved after reducing the number of cells used in the state
+# selector.
+
+# Count the max number of txs that the TxTable can hold by counting rows of
+# fixed fields + padding rows in the fixed fields section.
+def get_tx_table_max_txs(table: Set[TxTableRow]) -> int:
+    fixed_field_count = 0
+    for row in table:
+        print(vars(row))
+        if (row.field_tag != TxContextFieldTag.CallData) or (
+            row.field_tag == TxContextFieldTag.Pad and row.tx_id != 0
+        ):
+            fixed_field_count += 1
+    print(f"fixed_field_count = {fixed_field_count}")
+    return fixed_field_count // TxContextFieldTag.TxSignHash
 
 
 def end_block(instruction: Instruction):
     if instruction.is_last_step:
-        # Verify final step has tx_id identical to the tx amount in tx_table.
-        total_tx = instruction.call_context_lookup(CallContextFieldTag.TxId)
-        instruction.constrain_equal(
-            total_tx,
-            FQ(max([row.tx_id.expr().n for row in instruction.tables.tx_table])),
-        )
-
-        # Verify rw_counter counts to identical rw amount in rw_table to ensure
-        # there is no malicious insertion.
+        # 1. Verify rw_counter counts to the same number of meaningful rows in
+        # rw_table to ensure there is no malicious insertion.
+        rw_table_size = len(instruction.tables.rw_table)
         total_rw = instruction.curr.rw_counter + 1  # extra 1 from the tx_id lookup
-        instruction.constrain_equal(
-            total_rw,
-            FQ(len(instruction.tables.rw_table)),
+        # Verify that there are at most total_rw meaningful entries in the rw_table
+        instruction.rw_table_start_lookup(FQ(1))
+        instruction.rw_table_start_lookup(rw_table_size - total_rw)
+        # Since every lookup done in the EVM circuit must succeed and uses a unique
+        # rw_counter, we know that at least there are total_rw meaningful entries
+        # in the rw_table.
+        # We conclude that the number of meaningful entries in the rw_table is total_rw.
+
+        # 2. Verify that final step as tx_id identical to the number of txs in
+        # tx_table.
+        tx_table_max_txs = get_tx_table_max_txs(instruction.tables.tx_table)
+        total_tx = instruction.call_context_lookup(CallContextFieldTag.TxId)
+        # Verify that there are at most total_txs meaningful entries in the tx_table
+        instruction.tx_context_lookup(FQ(1), TxContextFieldTag.Pad)
+        instruction.tx_context_lookup(
+            FQ((tx_table_max_txs - total_tx.expr().n) * TxContextFieldTag.TxSignHash),
+            TxContextFieldTag.Pad,
         )
+        # Since every tx lookup done in the EVM circuit must succeed and uses a unique
+        # tx_id, we know that at least there are total_tx meaningful txs
+        # in the tx_table.
+        # We conclude that the number of meaningful txs in the tx_table is total_tx.
```

Comment:
With padding it seems to add more constraints to construct the `tx_table`, so rather than enforcing rows with tag `Pad` in `tx_table` to have sequential ids, is that possible to have a `NumTx` tag somewhere (perhaps `tx_table` with `tx_id = 0` row?) for us to check directly?

Reply:
Currently the `tx_table` is constructed by the verifier (out of publicly available data), so we don't have any constraint on it.  So having the sequential ids in the `Pad` rows doesn't need any constraint, it's just a logic that the verifier must follow when preparing the table contents.
A part from this, I think I don't understand your proposal.  Do you mean to have a tag that appears only in one row which contains the total number of Txs?  If we consider that the `tx_table` is constructed by the verifier, that seems good to me.  But if this proposal is to reduce the constraint complexity, I don't see why this constraint (which requires counting txs, which I imagine would need a new column) is simpler.

Reply:
If verifier always calculate same length RLC then it makes sense to me now, I thought we'd need constraint for that inside circuit, sorry for confusing.

What I imagined is that verifier doesn't need always to calculate so same length RLC when there are fewer txs than the maximum, the rest values should be auto padded with `0` because we don't add any value into RLC (or we should pad with sequential ids? otherwise prover could pad `0` ahead in advice column and still get the same RLC value, will think about this again). So instead we add another tag `NumTx` and put the value within such row, then we can lookup it directly.

Reply:
Ah I see, so you're thinking about having trailing 0s in the RLC'd array so that you can skip the operations for the tail of the array.  The idea makes sense!  Nevertheless the current arrangement of the tx table into the RLC'd value is column after column, so you would get the trailing 0s in the middle.

I think this is an interesting discussion so I'll move it to an issue so that we can move on with this PR :)

Reply:
Sure! Let's merge this PR first :)

Reply:
Moved to https://github.com/privacy-scaling-explorations/zkevm-specs/issues/253